wikipedia_cleanup/random_forest.py

```diff
@@ -0,0 +1,141 @@
+import hashlib
+import pickle
+from datetime import date, datetime, timedelta
+from pathlib import Path
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+from sklearn.ensemble import RandomForestClassifier
+from tqdm.auto import tqdm
+
+from wikipedia_cleanup.predictor import Predictor
+
+
+class RandomForestPredictor(Predictor):
+    def __init__(self, use_cache: bool = True) -> None:
+        super().__init__()
+        self.regressors: dict = {}
+        self.last_preds: dict = {}
+        self.hash_location = Path("cache") / self.__class__.__name__
+        self.use_hash = use_cache
+
+    @staticmethod
+    def get_relevant_ids(identifier: Tuple) -> List[Tuple]:
+        return []
+
+    @staticmethod
+    def get_relevant_attributes() -> List[str]:
+        return [
+            "value_valid_from",
+            "day_of_year",  # values from feature engineering
+            "day_of_month",
+            "day_of_week",
+            "month_of_year",
+            "quarter_of_year",
+            "is_month_start",
+            "is_month_end",
+            "is_quarter_start",
+            "is_quarter_end",
+            "days_since_last_change",
+            "days_since_last_2_changes",
+            "days_since_last_3_changes",
+            "days_between_last_and_2nd_to_last_change",
+            "days_until_next_change",
+            "mean_change_frequency_all_previous",
+            # check if this really improves the prediction
+            "mean_change_frequency_last_3",
+        ]
+
+    def _load_cache(self, possible_cached_mapping: Path) -> bool:
+        try:
+            if possible_cached_mapping.exists():
+                print(f"Cached model found, loading from {possible_cached_mapping}")
+                with open(possible_cached_mapping, "rb") as f:
+                    self.related_properties_lookup = pickle.load(f)
+                return True
+            else:
+                print("No cache found, recalculating model.")
+        except EOFError:
+            print("Caching failed, recalculating model.")
+        return False
+
+    def fit(
+        self, train_data: pd.DataFrame, last_day: datetime, keys: List[str]
+    ) -> None:
+        if self.use_hash:
+            possible_cached_mapping = self._calculate_cache_name(train_data)
+            if self._load_cache(possible_cached_mapping):
+                return
+        ilocs = train_data.groupby(keys, sort=False).count()["value_valid_from"]
```

Comment:
`itertools.groupby` is extremely fast on sorted lists and i think does roughly what you are emulating with the `iloc` variable. We used in in `predict.py`